libsplinter/src/admin/lifecycle/orchestrator.rs

```diff
@@ -0,0 +1,168 @@
+            service_type: service_type.to_string(),
+        };
+
+        let mut arg_map = HashMap::new();
+        for (key, value) in args {
+            arg_map.insert(key, value);
+        }
+
+        self.initialize_service(service_defination, arg_map)
+            .map_err(|err| InternalError::from_source(Box::new(err)))
+    }
+
+    fn retire_service(
+        &self,
+        circuit_id: &str,
+        service_id: &str,
+        service_type: &str,
+    ) -> Result<(), InternalError> {
+        if !self
+            .supported_service_types()
+            .contains(&service_type.to_string())
+        {
+            trace!(
+                "Ignoring call to add service, service type not supported: {}",
+                service_type
+            );
+            return Ok(());
+        }
+
+        debug!(
+            "Retire service: {}::{} ({})",
+            circuit_id, service_id, service_type,
+        );
+
+        let service_defination = ServiceDefinition {
+            circuit: circuit_id.to_string(),
+            service_id: service_id.to_string(),
+            service_type: service_type.to_string(),
+        };
+
+        self.stop_service(&service_defination)
+            .map_err(|err| InternalError::from_source(Box::new(err)))
+    }
+
+    fn purge_service(
+        &self,
+        circuit_id: &str,
+        service_id: &str,
+        service_type: &str,
+    ) -> Result<(), InternalError> {
+        if !self
+            .supported_service_types()
+            .contains(&service_type.to_string())
+        {
+            trace!(
+                "Ignoring call to add service, service type not supported: {}",
+                service_type
+            );
+            return Ok(());
+        }
+
+        debug!(
+            "Purge service: {}::{} ({})",
+            circuit_id, service_id, service_type,
+        );
+
+        let service_defination = ServiceDefinition {
+            circuit: circuit_id.to_string(),
+            service_id: service_id.to_string(),
+            service_type: service_type.to_string(),
+        };
+
+        self.purge_service(&service_defination)
+            .map_err(|err| InternalError::from_source(Box::new(err)))
+    }
+
+    fn shutdown_all_services(&self) -> Result<(), InternalError> {
+        debug!("Shutdown all services");
+        self.shutdown_all_services()
+            .map_err(|err| InternalError::from_source(Box::new(err)))
+    }
+
+    fn add_stopped_service(
+        &self,
+        circuit_id: &str,
+        service_id: &str,
+        service_type: &str,
+        args: HashMap<String, String>,
+    ) -> Result<(), InternalError> {
+        if !self
+            .supported_service_types()
+            .contains(&service_type.to_string())
+        {
+            trace!(
+                "Ignoring call to add service, service type not supported: {}",
+                service_type
+            );
+            return Ok(());
+        }
+
+        debug!(
```

Comment:
Does this double any logging in the existing `add_stopped_service` method?

Reply:
Nope, there is no logging that I was able to find.